src/openrct2/actions/StaffHireNewAction.hpp

```diff
@@ -264,48 +260,63 @@ DEFINE_GAME_ACTION(StaffHireNewAction, GAME_COMMAND_HIRE_NEW_STAFF_MEMBER, Staff
     void AutoPositionNewStaff(Peep * newPeep) const
     {
         // Find a location to place new staff member
-
         newPeep->State = PEEP_STATE_FALLING;
 
-        int16_t x, y, z;
         uint32_t count = 0;
-        uint16_t sprite_index;
-        Peep* guest = nullptr;
         PathElement* guest_tile = nullptr;
 
         // Count number of walking guests
-        FOR_ALL_GUESTS (sprite_index, guest)
         {
-            if (guest->State == PEEP_STATE_WALKING)
+            for (auto guest : EntityList<Guest>(SPRITE_LIST_PEEP))
             {
-                // Check the walking guest's tile. Only count them if they're on a path tile.
-                guest_tile = map_get_path_element_at(TileCoordsXYZ{ guest->NextLoc });
-                if (guest_tile != nullptr)
-                    ++count;
+                if (guest->State == PEEP_STATE_WALKING)
+                {
+                    // Check the walking guest's tile. Only count them if they're on a path tile.
+                    guest_tile = map_get_path_element_at(TileCoordsXYZ{ guest->NextLoc });
+                    if (guest_tile != nullptr)
+                        ++count;
+                }
             }
         }
 
+        CoordsXYZ newLocation{};
         if (count > 0)
         {
             // Place staff at a random guest
             uint32_t rand = scenario_rand_max(count);
-            FOR_ALL_GUESTS (sprite_index, guest)
+            Guest* chosenGuest = nullptr;
+
+            for (auto guest : EntityList<Guest>(SPRITE_LIST_PEEP))
             {
                 if (guest->State == PEEP_STATE_WALKING)
                 {
                     guest_tile = map_get_path_element_at(TileCoordsXYZ{ guest->NextLoc });
                     if (guest_tile != nullptr)
                     {
                         if (rand == 0)
+                        {
+                            chosenGuest = guest;
                             break;
+                        }
                         --rand;
                     }
                 }
             }
 
-            x = guest->x;
-            y = guest->y;
-            z = guest->z;
+            if (chosenGuest != nullptr)
+            {
+                newLocation.x = chosenGuest->x;
+                newLocation.y = chosenGuest->y;
+                newLocation.z = chosenGuest->z;
+            }
+            else
+            {
+                // User must pick a location
+                newPeep->State = PEEP_STATE_PICKED;
+                newLocation.x = newPeep->x;
+                newLocation.y = newPeep->y;
+                newLocation.z = newPeep->z;
```

Comment:
Could you explain these lines?

Reply:
If the lookup fails. This will revert to pickup state. Before it would crash on a nullptr exception